Crash during linking SDK 8.0

I just upgraded to SDK 8.0 (I'm not sure exactly which one I was using previously).
Compilation seemed to go OK, but the linking caused a crash:

```
mos-cx16-clang -Icx16 -Os  build_cx16/cx16/sys_cx16.o build_cx16/game.o build_cx16/gob.o build_cx16/player.o build_cx16/highscore.o build_cx16/input.o build_cx16/misc.o build_cx16/bub.o build_cx16/sfx.o build_cx16/story.o build_cx16/vfx.o build_cx16/export_cx16/palette.o build_cx16/cx16/irq.o  -o build_cx16/game.prg
PLEASE submit a bug report to https://github.com/llvm/llvm-project/issues/ and include the crash backtrace.
Stack dump:
0.	Program arguments: /home/ben/local/llvm-mos/bin/ld.lld build_cx16/cx16/sys_cx16.o build_cx16/game.o build_cx16/gob.o build_cx16/player.o build_cx16/highscore.o build_cx16/input.o build_cx16/misc.o build_cx16/bub.o build_cx16/sfx.o build_cx16/story.o build_cx16/vfx.o build_cx16/export_cx16/palette.o build_cx16/cx16/irq.o -plugin-opt=mcpu=mosw65c02 -plugin-opt=O2 -mllvm -force-precise-rotation-cost -mllvm -jump-inst-cost=6 -mllvm -force-loop-cold-block -mllvm -phi-node-folding-threshold=0 -mllvm -speculate-blocks=0 -mllvm -align-large-globals=false -mllvm -disable-spill-hoist -mllvm -lsr-complexity-limit=10000000 -mllvm -zp-avail=90 --gc-sections --sort-section=alignment -L/home/ben/local/llvm-mos/bin/../mos-platform/cx16/lib -L/home/ben/local/llvm-mos/bin/../mos-platform/commodore/lib -L/home/ben/local/llvm-mos/bin/../mos-platform/common/lib -l:crt0.o -lcrt0 -lcrt -lc -Tlink.ld -o build_cx16/game.prg
1.	Running pass 'Function Pass Manager' on module 'ld-temp.o'.
2.	Running pass 'MOSCombiner' on function '@main'
 #0 0x0000564300e566f7 llvm::sys::PrintStackTrace(llvm::raw_ostream&, int) (/home/ben/local/llvm-mos/bin/ld.lld+0x6976f7)
 llvm-mos/llvm-mos-sdk#1 0x0000564300e54f29 llvm::sys::RunSignalHandlers() (/home/ben/local/llvm-mos/bin/ld.lld+0x695f29)
 llvm-mos/llvm-mos-sdk#2 0x0000564300e55af6 (/home/ben/local/llvm-mos/bin/ld.lld+0x696af6)
 llvm-mos/llvm-mos-sdk#3 0x00007fd012d7b510 (/lib/x86_64-linux-gnu/libc.so.6+0x3c510)
 llvm-mos/llvm-mos-sdk#4 0x00005643011e0538 llvm::isConstantOrConstantSplatVector(llvm::MachineInstr&, llvm::MachineRegisterInfo const&) (/home/ben/local/llvm-mos/bin/ld.lld+0xa21538)
 llvm-mos/llvm-mos-sdk#5 0x0000564301178d96 llvm::CombinerHelper::matchConstantOp(llvm::MachineOperand const&, long) (/home/ben/local/llvm-mos/bin/ld.lld+0x9b9d96)
 llvm-mos/llvm-mos-sdk#6 0x00005643010c607b (/home/ben/local/llvm-mos/bin/ld.lld+0x90707b)
 llvm-mos/llvm-mos-sdk#7 0x00005643010c9fd8 (/home/ben/local/llvm-mos/bin/ld.lld+0x90afd8)
 llvm-mos/llvm-mos-sdk#8 0x00005643010cb9ff (/home/ben/local/llvm-mos/bin/ld.lld+0x90c9ff)
 llvm-mos/llvm-mos-sdk#9 0x000056430116b79e llvm::Combiner::combineMachineInstrs(llvm::MachineFunction&, llvm::GISelCSEInfo*) (/home/ben/local/llvm-mos/bin/ld.lld+0x9ac79e)
llvm-mos/llvm-mos-sdk#10 0x00005643010c5066 (/home/ben/local/llvm-mos/bin/ld.lld+0x906066)
llvm-mos/llvm-mos-sdk#11 0x00005643014b4e67 (/home/ben/local/llvm-mos/bin/ld.lld+0xcf5e67)
llvm-mos/llvm-mos-sdk#12 0x000056430209e5ec llvm::FPPassManager::runOnFunction(llvm::Function&) (/home/ben/local/llvm-mos/bin/ld.lld+0x18df5ec)
llvm-mos/llvm-mos-sdk#13 0x000056430209e759 llvm::FPPassManager::runOnModule(llvm::Module&) (/home/ben/local/llvm-mos/bin/ld.lld+0x18df759)
llvm-mos/llvm-mos-sdk#14 0x000056430209ee1a llvm::legacy::PassManagerImpl::run(llvm::Module&) (/home/ben/local/llvm-mos/bin/ld.lld+0x18dfe1a)
llvm-mos/llvm-mos-sdk#15 0x00005643013b20b1 (/home/ben/local/llvm-mos/bin/ld.lld+0xbf30b1)
llvm-mos/llvm-mos-sdk#16 0x00005643013b225b llvm::lto::backend(llvm::lto::Config const&, std::function<llvm::Expected<std::unique_ptr<llvm::CachedFileStream, std::default_delete<llvm::CachedFileStream>>> (unsigned int, llvm::Twine const&)>, unsigned int, llvm::Module&, llvm::ModuleSummaryIndex&) (/home/ben/local/llvm-mos/bin/ld.lld+0xbf325b)
llvm-mos/llvm-mos-sdk#17 0x00005643013ab311 llvm::lto::LTO::runRegularLTO(std::function<llvm::Expected<std::unique_ptr<llvm::CachedFileStream, std::default_delete<llvm::CachedFileStream>>> (unsigned int, llvm::Twine const&)>) (/home/ben/local/llvm-mos/bin/ld.lld+0xbec311)
llvm-mos/llvm-mos-sdk#18 0x00005643013ab607 llvm::lto::LTO::run(std::function<llvm::Expected<std::unique_ptr<llvm::CachedFileStream, std::default_delete<llvm::CachedFileStream>>> (unsigned int, llvm::Twine const&)>, std::function<llvm::Expected<std::function<llvm::Expected<std::unique_ptr<llvm::CachedFileStream, std::default_delete<llvm::CachedFileStream>>> (unsigned int, llvm::Twine const&)>> (unsigned int, llvm::StringRef, llvm::Twine const&)>) (/home/ben/local/llvm-mos/bin/ld.lld+0xbec607)
llvm-mos/llvm-mos-sdk#19 0x0000564300f4b9c6 lld::elf::BitcodeCompiler::compile() (/home/ben/local/llvm-mos/bin/ld.lld+0x78c9c6)
llvm-mos/llvm-mos-sdk#20 0x0000564300ef1218 void lld::elf::LinkerDriver::compileBitcodeFiles<llvm::object::ELFType<(llvm::support::endianness)1, false>>(bool) (/home/ben/local/llvm-mos/bin/ld.lld+0x732218)
llvm-mos/llvm-mos-sdk#21 0x0000564300ef8791 lld::elf::LinkerDriver::link(llvm::opt::InputArgList&) (/home/ben/local/llvm-mos/bin/ld.lld+0x739791)
llvm-mos/llvm-mos-sdk#22 0x0000564300f00a2f lld::elf::LinkerDriver::linkerMain(llvm::ArrayRef<char const*>) (/home/ben/local/llvm-mos/bin/ld.lld+0x741a2f)
llvm-mos/llvm-mos-sdk#23 0x0000564300f0192c lld::elf::link(llvm::ArrayRef<char const*>, llvm::raw_ostream&, llvm::raw_ostream&, bool, bool) (/home/ben/local/llvm-mos/bin/ld.lld+0x74292c)
llvm-mos/llvm-mos-sdk#24 0x0000564300e7156d lld::unsafeLldMain(llvm::ArrayRef<char const*>, llvm::raw_ostream&, llvm::raw_ostream&, llvm::ArrayRef<lld::DriverDef>, bool) (/home/ben/local/llvm-mos/bin/ld.lld+0x6b256d)
llvm-mos/llvm-mos-sdk#25 0x0000564300e12f03 lld_main(int, char**, llvm::ToolContext const&) (/home/ben/local/llvm-mos/bin/ld.lld+0x653f03)
llvm-mos/llvm-mos-sdk#26 0x0000564300ceab86 main (/home/ben/local/llvm-mos/bin/ld.lld+0x52bb86)
llvm-mos/llvm-mos-sdk#27 0x00007fd012d666ca (/lib/x86_64-linux-gnu/libc.so.6+0x276ca)
llvm-mos/llvm-mos-sdk#28 0x00007fd012d66785 __libc_start_main (/lib/x86_64-linux-gnu/libc.so.6+0x27785)
llvm-mos/llvm-mos-sdk#29 0x0000564300e12b1e _start (/home/ben/local/llvm-mos/bin/ld.lld+0x653b1e)
mos-cx16-clang: error: unable to execute command: Segmentation fault
mos-cx16-clang: error: ld.lld command failed due to signal (use -v to see invocation)
make: *** [Makefile.cx16:45: build_cx16/game.prg] Error 1
```

further info:
My code _was_ just creeping over the 38KB limit for the cx16 ($0800 - $9EFF).
Under the older SDK, linking would fail (without crashing):
```
mos-cx16-clang -Icx16 -Os  build_cx16/cx16/sys_cx16.o build_cx16/game.o build_cx16/gob.o build_cx16/player.o build_cx16/highscore.o build_cx16/input.o build_cx16/misc.o build_cx16/bub.o build_cx16/sfx.o build_cx16/story.o build_cx16/vfx.o build_cx16/export_cx16/palette.o build_cx16/cx16/irq.o  -o build_cx16/game.prg
ld.lld: error: section '.bss' will not fit in region 'ram': overflowed by 324 bytes
ld.lld: error: section '.noinit' will not fit in region 'ram': overflowed by 340 bytes
mos-cx16-clang: error: ld.lld command failed with exit code 1 (use -v to see invocation)
make: *** [Makefile.cx16:45: build_cx16/game.prg] Error 1
```

(I wanted to see if the new SDK produced more compact code, so I could just sneak in under that 38KB limit).

--- a/llvm/lib/Target/MOS/MOSCombine.td
+++ b/llvm/lib/Target/MOS/MOSCombine.td
@@ -101,5 +101,5 @@ def MOSCombinerHelper : GICombinerHelper<"MOSCombinerImpl", [
   sbc_equal,
   fold_shift,
   shift_unused_carry_in,
-  mul_to_shift_and_add
+  mul_to_shift_and_add,
 ]>;

--- a/llvm/lib/Target/MOS/MOSCombiner.cpp
+++ b/llvm/lib/Target/MOS/MOSCombiner.cpp
@@ -638,9 +638,7 @@ bool MOSCombinerImpl::matchMulToShiftAndAdd(MachineInstr &MI,
                  : B.buildInstr(MathOp, {Ty}, {Shl, LHS});
     if (RHSConst.isNegative())
       R = B.buildNeg(Ty, R);
-    Observer.changingInstr(*R);
-    R->getOperand(0).setReg(MI.getOperand(0).getReg());
-    Observer.changedInstr(*R);
+    B.buildCopy(MI.getOperand(0).getReg(), R);
     MI.eraseFromParent();
   };
   return true;
@@ -681,7 +679,8 @@ APInt MOSCombinerImpl::getDemandedBits(Register R,
     case MOS::G_LSHRE: {
       APInt DstDemandedBits = getDemandedBits(MI.getOperand(0).getReg(), Cache);
       if (Use.getOperandNo() == 2) {
-        APInt CarryOutDemanded = getDemandedBits(MI.getOperand(1).getReg(), Cache);
+        APInt CarryOutDemanded =
+            getDemandedBits(MI.getOperand(1).getReg(), Cache);
         DemandedBits |= DstDemandedBits << 1 | CarryOutDemanded.zext(8);
       } else {
         assert(Use.getOperandNo() == 3);
@@ -692,7 +691,8 @@ APInt MOSCombinerImpl::getDemandedBits(Register R,
     case MOS::G_SHLE: {
       APInt DstDemandedBits = getDemandedBits(MI.getOperand(0).getReg(), Cache);
       if (Use.getOperandNo() == 2) {
-        APInt CarryOutDemanded = getDemandedBits(MI.getOperand(1).getReg(), Cache);
+        APInt CarryOutDemanded =
+            getDemandedBits(MI.getOperand(1).getReg(), Cache);
         DemandedBits |=
             DstDemandedBits.lshr(1) | (CarryOutDemanded.zext(8) << 7);
       } else {

